components/provisioner/internal/installation/resource_cleanup.go

```diff
@@ -0,0 +1,148 @@
+package installation
+
+import (
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/kubernetes-sigs/service-catalog/pkg/apis/servicecatalog/v1beta1"
+	sc "github.com/kubernetes-sigs/service-catalog/pkg/client/clientset_generated/clientset"
+	"github.com/pkg/errors"
+	apiErrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/rest"
+	"k8s.io/klog"
+)
+
+const (
+	BrokerUrlPrefix                      = "https://service-manager."
+	ClusterServiceBrokerNameLabel        = "servicecatalog.k8s.io/spec.clusterServiceBrokerName"
+	ClusterServiceClassExternalNameLabel = "servicecatalog.k8s.io/spec.externalName"
+	ClusterServiceClassRefNameLabel      = "servicecatalog.k8s.io/spec.clusterServiceClassRef.name"
+)
+
+type ResourceClient interface {
+	PerformCleanup() error
+}
+
+type ServiceCatalogClient interface {
+	PerformCleanup() error
+	ListClusterServiceBroker(metav1.ListOptions) (*v1beta1.ClusterServiceBrokerList, error)
+	ListClusterServiceClass(metav1.ListOptions) (*v1beta1.ClusterServiceClassList, error)
+	ListServiceInstance(metav1.ListOptions) (*v1beta1.ServiceInstanceList, error)
+}
+
+func NewServiceCatalogClient(kubeconfig *rest.Config) (ServiceCatalogClient, error) {
+	scCli, err := sc.NewForConfig(kubeconfig)
+	if err != nil {
+		return &serviceCatalogClient{}, err
+	}
+
+	return &serviceCatalogClient{client: scCli}, nil
+}
+
+type serviceCatalogClient struct {
+	client sc.Interface
+}
+
+func (s *serviceCatalogClient) PerformCleanup() error {
```

Comment:
Please consider splitting this function into smaller ones.

Reply:
done